### ISSUE TITLE
Build and Dev not working after setting up a Yarn 2.0 monorepo with PnP and TypeSript

### ISSUE THREAD

**Issue by dahaca** [vercel/next.js#27629]
### What version of Next.js are you using?

11.0.1

### What version of Node.js are you using?

14.17.3

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

next start

### Describe the Bug

Recently I began transitioning my project to a Yarn 2 + PnP monorepo. I did it in steps.

- [X] Move to Yarn 2 + `node_modules`
- [X] Move to PnP
- [ ] Move to the monorepo architecture

Everything was working as a single project repository. I installed all needed SDKs for VSCode, had to additionally install `events` and `string_decoder` polyfills as webpack 5 was suddenly missing those. Also `eslint-import-resolver-node`. Otherwise, everything was left as is and both `dev` and `build` + `start` were working without any problems!

Until I started moving folders around. Suddenly, nothing works, modules cannot be found. I was following many examples of Yarn 2 monorepos, monorepos with PnP, and Next and cannot understand why modules are suddenly perceived as missing.v  know it is likely not an issue of Next but perhaps somebody would at least know what's up 😀

When running `dev` I get that:

```
error - ../../.yarn/__virtual__/redux-persist-virtual-8233dba7d6/0/cache/redux-persist-npm-6.0.0-56dad69feb-edaf10dbf1.zip/node_modules/redux-persist/es/integration/react.js:21:0
Module not found: Can't resolve 'react'
Did you mean './react'?
Requests that should resolve in the current directory need to start with './'.
Requests that start with a name are treated as module requests and resolve within module directories (node_modules, /home/node/cl/packages/cl).
If changing the source code is not an option there is also a resolve option called 'preferRelative' which tries to resolve these kinds of requests in the current directory too.
```

While `build` fails with that:

```
> Build error occurred
Error: Failed to load parser '@typescript-eslint/parser' declared in '.eslintrc » ../../.eslintrc.base.json': Your application tried to access @typescript-eslint/parser, but it isn't declared in your dependencies; this makes the require call ambiguous and unsound.

Required package: @typescript-eslint/parser
Required by: /home/node/cl/

Require stack:
- /home/node/cl/.eslintrc.base.json
    at Function.external_module_.Module._resolveFilename (/home/node/cl/.pnp.cjs:24850:55)
    at Function.resolve (internal/modules/cjs/helpers.js:98:19)
    at Object.resolve (/home/node/cl/.yarn/cache/@eslint-eslintrc-npm-0.4.3-ee1bbcab87-03a7704150.zip/node_modules/@eslint/eslintrc/lib/shared/relative-module-resolver.js:28:50)
    at ConfigArrayFactory._loadParser (/home/node/cl/.yarn/cache/@eslint-eslintrc-npm-0.4.3-ee1bbcab87-03a7704150.zip/node_modules/@eslint/eslintrc/lib/config-array-factory.js:929:45)
    at ConfigArrayFactory._normalizeObjectConfigDataBody (/home/node/cl/.yarn/cache/@eslint-eslintrc-npm-0.4.3-ee1bbcab87-03a7704150.zip/node_modules/@eslint/eslintrc/lib/config-array-factory.js:724:43)
    at _normalizeObjectConfigDataBody.next (<anonymous>)
    at ConfigArrayFactory._normalizeObjectConfigData (/home/node/cl/.yarn/cache/@eslint-eslintrc-npm-0.4.3-ee1bbcab87-03a7704150.zip/node_modules/@eslint/eslintrc/lib/config-array-factory.js:665:20)
    at _normalizeObjectConfigData.next (<anonymous>)
    at ConfigArrayFactory._normalizeObjectConfigDataBody (/home/node/cl/.yarn/cache/@eslint-eslintrc-npm-0.4.3-ee1bbcab87-03a7704150.zip/node_modules/@eslint/eslintrc/lib/config-array-factory.js:720:25)
    at _normalizeObjectConfigDataBody.next (<anonymous>) {
  type: 'Error'
}
```

In both cases, the modules are of course installed.

### Expected Behavior

`dev` and `start` scripts run as they should

### To Reproduce

Here is the current file structure:

```
app/
├─ .yarn/
├─ packages/
│  ├─ cl/
│  │  ├─ .babelrc
│  │  ├─ .eslintrc
│  │  ├─ next.config.js
│  │  ├─ package.json
│  │  ├─ tsconfig.json
├─ .eslintrc.base.json
├─ .pnp.cjs
├─ .yarnrc.yml
├─ package.json
├─ tsconfig.base.json
├─ yarn.lock
```

Root `package.json` (the scripts are temporary):

```
{
  "name": "monorepo",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "yarn workspace cl dev",
    "build": "yarn workspace cl build",
    "analyze": "yarn workspace cl analyze",
    "start": "yarn workspace cl start",
    "compile": "yarn workspace cl compile",
    "lint": "yarn workspace cl lint",
    "test": "yarn workspace cl test",
    "cypress": "yarn workspace cl cypress",
    "e2e": "yarn workspace cl e2e",
    "prepare": "yarn workspace cl prepare"
  },
  "workspaces": [
    "packages/*"
  ],
  "devDependencies": {
    "husky": "^7.0.1",
    "prettier": "^2.3.2",
    "pretty-quick": "^3.1.1"
  }
}
```

```
{
  "name": "cl",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "analyze": "ANALYZE=true yarn build",
    "start": "next start",
    "compile": "tsc -p tsconfig.json",
    "lint": "eslint --ext=js,ts,tsx --fix .",
    "test": "jest --watchAll",
    "cypress": "cypress open",
    "e2e": "cypress run",
    "prepare": "husky install"
  },
  "dependencies": {
   *****
    "app": "link:./",
  },
  "devDependencies": {
   ******
    "typescript": "^4.3.5"
  }
}

```

Base `tsconfig.base.json`:

```
{
  "compilerOptions": {
    "target": "es2018",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": false,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "noUnusedParameters": true,
    "noUnusedLocals": true,
    "noImplicitAny": true
  }
}
```

```
{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "baseUrl": "./",
    "paths": {
      "./*": ["app/*"]
    }
  },
  "exclude": ["cypress"],
  "include": [
    "next-env.d.ts",
    "**/*.ts",
    "**/*.tsx",
    "**/*.svg",
    "next-i18next.config.js"
  ]
}
```

For better DX I used the `link:` [protocol](https://yarnpkg.com/features/protocols#why-is-the-link-protocol-recommended-over-aliases-for-path-mapping) in the package's package.json and then mapped it back to `./` in tsconfig.json. Auto-imports work and there is no need to prepend local imports with `app/` – `utils/sort.ts` instead of `app/utils/sort.ts`. Absolute imports were not working without `link:` and auto-imports without `baseUrl`

**Comment by jonasgloning** [comment]
Hey @dahaca,

can you post your full `package.json`s and your `.pnp.cjs`? Or better yet, publish your codebase minus all proprietary parts.

```bash
> Build error occurred
Error: Failed to load parser '@typescript-eslint/parser' declared in '.eslintrc » ../../.eslintrc.base.json': Your application tried to access @typescript-eslint/parser, but it isn't declared in your dependencies; this makes the require call ambiguous and unsound.

Required package: @typescript-eslint/parser
```

indicates that `@typescript-eslint/parser` isn't listed in your project's dependencies.

**Comment by dahaca** [comment]
Hi @jonasgloning !

Thank you for your reply! Want to report that apparently, the problem was in `redux-persist`, which hasn't been updated for a while. I set the `pnpMode` to `loose` and it all started to work. Still not exactly sure why it was not able to find `react` and if there is another way to fix the issue without rewriting the package or changing the `pnpMode`. Speaking of the parser - yes, I had to install all eslint plugins in both root and the next directory for it to work.
